### PR TITLE
Master services4 jem

### PR DESCRIPTION
**Pull request by jem-odoo**
https://www.odoo.com/web?debug#id=34703&action=333&active_id=965&model=project.task&view_type=form&menu_id=4720

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale_timesheet/models/account.py [hunk 1]
```diff
@@ -48,8 +67,40 @@ def write(self, values):
...
+            if task.billable_type == 'employee_rate':
+                map_entry = self.env['project.sale.line.employee.map'].search([('project_id', '=', task.project_id.id), ('employee_id', '=', employee.id)])
+                if map_entry:
+                    return map_entry.sale_line_id
+                if task.sale_line_id:
+                    return task.sale_line_id
+                return task.project_id.sale_line_id
+            elif task.billable_type == 'task_rate':
+                return task.sale_line_id
+            return task.sale_line_id  # fallback for employee_rate and normal flow for task_rate
```
Comment (RomainLibert) [inline addons/sale_timesheet/models/account.py]: Seems to me that this fallback will never be reached as you always return before it

File: addons/sale_timesheet/models/account.py [hunk 2]
```diff
@@ -48,8 +67,40 @@ def write(self, values):
     @api.model
     def _timesheet_preprocess(self, values):
         values = super(AccountAnalyticLine, self)._timesheet_preprocess(values)
-        # task implies so line
-        if 'task_id' in values:
+        # task implies so line (at create)
+        if 'task_id' in values and not values.get('so_line') and values.get('employee_id'):
```
Comment (RomainLibert) [inline addons/sale_timesheet/models/account.py]: Why do we check if we have an employee here?

File: addons/sale_timesheet/models/account.py [hunk 3]
```diff
@@ -36,6 +37,24 @@ def _compute_timesheet_invoice_type(self):
                         invoice_type = 'billable_fixed'
                 timesheet.timesheet_invoice_type = invoice_type
 
+    @api.onchange('employee_id')
+    def _onchange_task_id_employee_id(self):
+        if self.project_id:  # timesheet only
+            if self.task_id.billable_type == 'task_rate':
+                self.so_line = self.task_id.sale_line_id
+            elif self.task_id.billable_type == 'employee_rate':
+                sale_line = self._timesheet_determine_sale_line(self.task_id, self.employee_id)
```
Comment (RomainLibert) [inline addons/sale_timesheet/models/account.py]: Could be done in one line

File: addons/sale_timesheet/models/project.py [hunk 4]
```diff
@@ -70,13 +80,25 @@ def _get_default_partner(self):
...
         if self._context.get('default_parent_id'):
-            sale_line_id = self.env['project.task'].browse(self._context['default_parent_id']).sale_line_id.id
+            parent_task = self.env['project.task'].browse(self._context['default_parent_id'])
+            if parent_task.project_id.billable_type != 'no':
+                sale_line_id = parent_task.sale_line_id.id
         if not sale_line_id and self._context.get('default_project_id'):
-            sale_line_id = self.env['project.project'].browse(self._context['default_project_id']).sale_line_id.id
+            project = self.env['project.project'].browse(self.env.context['default_project_id'])
+            if project.billable_type != 'no':
+                return project.sale_line_id
```
[the remark this answers is not in the document]
Comment (jem-odoo) [reply addons/sale_timesheet/models/project.py]: The SO line field is only visible when the customer is set on the task. In a non-billable project, usually, there is no customer set (by default from the project), so setting the SO line will be invisible and bring confusion.

**Pull request merged**